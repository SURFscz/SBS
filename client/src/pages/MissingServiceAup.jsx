import React from "react";
import {withRouter} from "react-router-dom";
import I18n from "i18n-js";
import "./MissingServiceAup.scss";
import Button from "../components/Button";
import {serviceAupBulkCreate} from "../api";
import SpinnerField from "../components/redesign/SpinnerField";
import CollaborationAupAcceptance from "../components/CollaborationAupAcceptance";


class MissingServiceAup extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            agreed: false,
            loading: false
        };
    }

    agreeWith = () => {
        this.setState({loading: true});
        const {user, reloadMe} = this.props;
        serviceAupBulkCreate(user.services_without_aup).then(res => {
            const url = new URL(res.location);
            reloadMe({
                user: user,
                callback: () => {
                    this.props.history.push(url.pathname + url.search);
                }
            })
        });
    }

    render() {
        const {agreed, loading} = this.state;
        const {user} = this.props;
        if (loading) {
            return <SpinnerField/>;
        }
        const info = user.services_without_aup.length > 1 ? "infoMultiple" : "info";
        const services = user.services_without_aup.length > 1 ? "informationServiceMultiple" : "informationService";
        return (
            <div className="mod-missing-service-aup">
                <h1>{I18n.t("aup.service.title")}</h1>
                <p className="info">{I18n.t(`aup.service.missing.${info}`)}</p>
                {user.service_collaborations.map(collaboration => <div className="collaboration-detail">
                    <h2 dangerouslySetInnerHTML={{__html: I18n.t("aup.service.purposeOf", {name: collaboration.name})}}/>
                    <p>{collaboration.description}</p>
                </div>)}
                <h2>{I18n.t(`aup.service.${services}`)}</h2>
                <CollaborationAupAcceptance services={user.services_without_aup}
                                            disabled={!agreed}
                                            serviceEmails={user.service_emails}
                                            setDisabled={() => this.setState({agreed: !agreed})}/>
                <div className="actions">
                    <Button className="proceed" onClick={this.agreeWith} centralize={true}
                            txt={I18n.t("aup.onward")} disabled={!agreed}/>
                </div>
            </div>
        )
    }
}

export default withRouter(MissingServiceAup);