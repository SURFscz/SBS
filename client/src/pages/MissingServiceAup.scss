@import "../stylesheets/vars.scss";

.mod-missing-service-aup {
    max-width: $compact-medium;
    margin: 25px auto 40px auto;
    width: 100%;

    @media (max-width: 680px) {
        padding: 0 15px;
    }

    h1 {
        font-size: 32px;
        line-height: 36px;
        margin-bottom: 25px;
    }

    h2 {
        font-size: 24px;
        line-height: 28px;
        margin: 20px 0 15px 0;
        color: $primary-orange;
    }

    p.info {
        margin: 20px 0;
    }

    div.actions {
        margin: 25px 0;
    }

}