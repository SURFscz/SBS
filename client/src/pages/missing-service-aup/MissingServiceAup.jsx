import React from "react";
import {withRouter} from "react-router-dom";
import I18n from "../../locale/I18n";
import "./MissingServiceAup.scss";
import Button from "../../components/button/Button";
import {serviceAupBulkCreate} from "../../api";
import SpinnerField from "../../components/redesign/spinner-field/SpinnerField";
import CollaborationAupAcceptance from "../../components/collaboration-aup-acceptance/CollaborationAupAcceptance";
import DOMPurify from "dompurify";


class MissingServiceAup extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            agreed: false,
            loading: false
        };
    }

    agreeWith = () => {
        this.setState({loading: true});
        const {user, reloadMe} = this.props;
        serviceAupBulkCreate(user.services_without_aup).then(res => {
            const urlSearchParams = new URLSearchParams(window.location.search);
            let path = urlSearchParams.get("state");
            if (path) {
                //user was already logged in
                path = decodeURIComponent(path);
            } else if (res.location) {
                const url = new URL(res.location);
                path = url.pathname + url.search;
            } else {
                path = "/"
            }
            reloadMe(() => this.props.history.push(path));
        });
    }

    render() {
        const {agreed, loading} = this.state;
        const {user} = this.props;
        if (loading) {
            return <SpinnerField/>;
        }
        const info = user.services_without_aup.length > 1 ? "infoMultiple" : "info";
        const services = user.services_without_aup.length > 1 ? "informationServiceMultiple" : "informationService";
        return (
            <div className="mod-missing-service-aup">
                <h1>{I18n.t("aup.service.title")}</h1>
                <p className="info">{I18n.t(`aup.service.missing.${info}`)}</p>
                {(user.service_collaborations || []).map((collaboration, index) =>
                    <div className="collaboration-detail" key={index}>
                        <h3 dangerouslySetInnerHTML={{__html: DOMPurify.sanitize(I18n.t("aup.service.purposeOf", {name: collaboration.name}))}}/>
                        <p>{collaboration.description}</p>
                    </div>)}
                <h3>{I18n.t(`aup.service.${services}`)}</h3>
                <CollaborationAupAcceptance services={user.services_without_aup}
                                            disabled={!agreed}
                                            serviceEmails={user.service_emails}
                                            setDisabled={() => this.setState({agreed: !agreed})}/>
                <div className="actions">
                    <Button className="proceed"
                            onClick={this.agreeWith}
                            centralize={true}
                            txt={I18n.t("aup.onward")}
                            disabled={!agreed}/>
                </div>
            </div>
        )
    }
}

export default withRouter(MissingServiceAup);
