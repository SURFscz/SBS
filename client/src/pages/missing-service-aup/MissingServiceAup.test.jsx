import React from "react";
import {render, fireEvent, waitFor} from "@testing-library/react";
import MissingServiceAup from "./MissingServiceAup";
import {MemoryRouter} from "react-router-dom";

jest.mock("../../api", () => ({
    serviceAupBulkCreate: jest.fn(() => Promise.resolve({location: "/some/path"})),
}));

const mockUser = {
    services_without_aup: [
        {id: "1", name: "Service 1"}
    ],
    service_collaborations: [
        {name: "Collab 1", description: "Description 1"}
    ],
    service_emails: ["test@example.com"]
};

const mockReloadMe = jest.fn(cb => cb());
const mockHistory = {push: jest.fn()};

describe("MissingServiceAup", () => {
    it("renders and matches snapshot", () => {
        const {asFragment} = render(
            <MemoryRouter>
                <MissingServiceAup
                    user={mockUser}
                    reloadMe={mockReloadMe}
                    history={mockHistory}
                />
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
