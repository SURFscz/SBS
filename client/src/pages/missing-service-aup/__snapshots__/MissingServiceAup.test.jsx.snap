// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissingServiceAup renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mod-missing-service-aup"
  >
    <h1>
      Review the information in order to continue
    </h1>
    <p
      class="info"
    >
      The application acceptable use policy (AUP) has changed. Please read it and indicate below whether you accept.
    </p>
    <div
      class="collaboration-detail"
    >
      <h3>
        Purpose of 
        <strong>
          Collab 1
        </strong>
      </h3>
      <p>
        Description 1
      </p>
    </div>
    <h3>
      Information about the application
    </h3>
    <div
      class="mod-aup-services"
    >
      <div
        class="services"
      >
        <div
          class="service-section"
        >
          <table>
            <thead />
            <tbody>
              <tr>
                <td
                  class="logo"
                />
                <td
                  class="service-detail"
                >
                  <div
                    class="service"
                  >
                    <span
                      class="service-name"
                    >
                      Service 1
                    </span>
                    <div
                      class="policies"
                    >
                      <div
                        class="policy"
                      >
                        <span
                          class="no-link"
                        >
                          No AUP provided
                        </span>
                      </div>
                      <div
                        class="policy"
                      >
                        <span
                          class="no-link"
                        >
                          No privacy policy provided
                        </span>
                      </div>
                      <div
                        class="policy contact"
                      >
                        <span
                          class="no-link"
                        >
                          No contact provided
                        </span>
                      </div>
                    </div>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <div
      class="actions"
    >
      <button
        class="sds--btn sds--btn--primary  sds--btn--full   "
        disabled=""
        type="button"
      >
        <span
          class="textual"
        >
          Onwards
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
